extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-plugin-service-impl.ts

```diff
@@ -0,0 +1,739 @@
+import { K8sDevWorkspaceEnvVariables } from './k8s-devworkspace-env-variables';
+import { PluginServiceImpl } from '@eclipse-che/theia-remote-api/lib/common/plugin-service-impl';
+
+export const CHE_PLUGINS_JSON = '/plugins/che-plugins.json';
+
+export const PLUGINS_DIR = '/plugins';
+export const SIDECAR_PLUGINS_DIR = '/plugins/sidecars';
+
+export interface PluginMetaYaml {
+  apiVersion?: string;
+  publisher?: string;
+  name?: string;
+  version?: string;
+  type?: string;
+  displayName?: string;
+  title?: string;
+}
+
+interface PluginList {
+  plugins: string[];
+}
+
+export class Deferred<T> {
+  state: 'pending' | 'resolved' | 'rejected' = 'pending';
+  resolve: (value?: T) => void;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  reject: (err?: any) => void;
+
+  promise = new Promise<T>((resolve, reject) => {
+    this.resolve = result => {
+      resolve(result as T);
+      if (this.state === 'pending') {
+        this.state = 'resolved';
+      }
+    };
+    this.reject = err => {
+      reject(err);
+      if (this.state === 'pending') {
+        this.state = 'rejected';
+      }
+    };
+  });
+}
+
+@injectable()
+export class K8sPluginServiceImpl extends PluginServiceImpl {
+  @inject(K8sDevWorkspaceEnvVariables)
+  private env: K8sDevWorkspaceEnvVariables;
+
+  @inject(DevfileService)
+  private devfileService: DevfileService;
+
+  deferred = new Deferred<void>();
+
+  @postConstruct()
+  protected startup(): void {
+    this.initialize()
+      .then(() => this.deferred.resolve())
+      .catch(e => this.deferred.reject(e));
+  }
+
+  async initialize(): Promise<void> {
+    const ERRMSG = 'Failure to initialize K8sPluginServiceImpl';
+
+    const projectsRoot = this.env.getProjectsRoot();
+
+    if (!projectsRoot) {
+      throw new Error(ERRMSG + 'Projects Root is not set.');
+    }
+
+    if (!(await fs.pathExists(projectsRoot))) {
+      throw new Error(ERRMSG + `Projects root [${projectsRoot}] does not exist`);
+    }
+
+    if (await fs.pathExists(CHE_PLUGINS_JSON)) {
+      return;
+    }
+
+    try {
+      const pluginList: PluginList = { plugins: [] };
+
+      const projects = await fs.readdir(projectsRoot);
+      for (const project of projects) {
+        const extensionsJson = path.join(projectsRoot, project, '.vscode', 'extensions.json');
+
+        if ((await fs.pathExists(extensionsJson)) && (await fs.stat(extensionsJson)).isFile()) {
+          try {
+            const content = await fs.readFile(extensionsJson, 'utf-8');
+
+            // remove c-style comments from JSON loke following
+            // {
+            //   // See https://go.microsoft.com/fwlink/?LinkId=827846
+            //   // for the documentation about the extensions.json format
+            //   "recommendations": [
+            //     "redhat.java"
+            //   ]
+            // }
+            const reg = /\/\/.*$/gm;
```

Comment:
FYI there is https://www.npmjs.com/package/jsonc-parser

Reply:
Done